sklearn/metrics/pairwise.py

```diff
@@ -1943,40 +1949,48 @@ def distance_metrics():
     return PAIRWISE_DISTANCE_FUNCTIONS
 
 
-def _dist_wrapper(dist_func, dist_matrix, slice_, *args, **kwargs):
+def _transposed_dist_wrapper(dist_func, dist_matrix, slice_, *args, **kwargs):
     """Write in-place to a slice of a distance matrix."""
-    dist_matrix[:, slice_] = dist_func(*args, **kwargs)
+    dist_matrix[slice_, ...] = dist_func(*args, **kwargs).T
 
 
 def _parallel_pairwise(X, Y, func, n_jobs, **kwds):
     """Break the pairwise matrix in n_jobs even slices
     and compute them using multithreading."""
+    xp, _, device = get_namespace_and_device(X, Y)
+    X, Y, dtype_float = _find_floating_dtype_allow_sparse(X, Y, xp=xp)
 
     if Y is None:
         Y = X
-    X, Y, dtype = _return_float_dtype(X, Y)
 
     if effective_n_jobs(n_jobs) == 1:
         return func(X, Y, **kwds)
 
     # enforce a threading backend to prevent data communication overhead
-    fd = delayed(_dist_wrapper)
-    ret = np.empty((X.shape[0], Y.shape[0]), dtype=dtype, order="F")
+    fd = delayed(_transposed_dist_wrapper)
+    # Transpose `ret` such that a given thread writes its ouput to a contiguous chunk.
+    # Note `order` (i.e. F/C-contiguous) is not included in array API standard, see
+    # https://github.com/data-apis/array-api/issues/571 for details.
+    # We assume that currently (April 2025) all array API compatible namespaces
+    # allocate 2D arrays using the C-contiguity convention by default.
+    ret = xp.empty((X.shape[0], Y.shape[0]), device=device, dtype=dtype_float).T
     Parallel(backend="threading", n_jobs=n_jobs)(
-        fd(func, ret, s, X, Y[s], **kwds)
+        fd(func, ret, s, X, Y[s, ...], **kwds)
         for s in gen_even_slices(_num_samples(Y), effective_n_jobs(n_jobs))
```

Comment:
> It would be interesting to do some benchmarking to assess the impact of the chunk size when using torch CUCDA or CuPy vs numpy CPU.

Just to confirm, we're talking about these chunk sizes right? :grimacing: